p2p/peering.go

```diff
@@ -819,6 +820,7 @@ func (rp *RemotePeer) receivedHeaders(ctx context.Context, msg *wire.MsgHeaders)
 		op := errors.Opf(opf, rp.raddr)
 		err := errors.E(op, errors.Protocol, "received unrequested headers")
 		rp.Disconnect(err)
+		rp.requestedHeadersMu.Unlock()
```

Comment:
this can be moved up to the top of the if block to reduce some lock contention

Reply:
Isn't it better to first do the `rp.Disconnect(err)` that will close the socket, then right after unlock the mutex?

Reply:
It doesn't actually matter in this case since Disconnect is lockless (instead using a CAS to check if the disconnect has already happened or is in progress), but if Disconnect was ever changed to use a mutex, that would result in nested held locks and we would need to define a locking order.  I can see the argument that this way is better since it does avoid a race of which error a peer is disconnected with.

Reply:
Got it, thanks 👍 I was not really sure about the behaviour of Disconnect since I just took a quick look on the code. Just let me know if you'd like to keep it like that or not.

Btw, I might be interested in working part time for Decred, seems like an interesting project 🙃